File: shifai-ios/ShifAI/Data/CrashReporter.swift
import Foundation

/// Crash Reporter — lightweight error tracking for production.
/// No third-party dependency: logs to local file + optional Plausible events.
/// Follows zero-PII principle: no stack traces with user data.
final class CrashReporter {

    static let shared = CrashReporter()

    private let logFile: URL
    private let maxLogSize = 500_000  // 500 KB
    private let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return f
    }()

    private init() {
        logFile = FileManager.default
            .containerURL(forSecurityApplicationGroupIdentifier: "group.com.shifai.shared")!
            .appendingPathComponent("crash_log.txt")
    }

    // MARK: - Logging

    func log(_ error: ShifAIError, context: String = "") {
        let timestamp = dateFormatter.string(from: Date())
        let line = "[\(timestamp)] [\(error.code)] \(error.localizedDescription) | \(context)\n"

        appendToFile(line)

        // Track anonymized error in analytics
        Task {
            await AnalyticsTracker.shared.trackError(errorCode: error.code)
        }
    }

    func logCrash(_ error: Error, file: String = #file, line: Int = #line) {
        let filename = URL(fileURLWithPath: file).lastPathComponent
        let entry = "[CRASH] \(dateFormatter.string(from: Date())) | \(filename):\(line) | \(error.localizedDescription)\n"
        appendToFile(entry)
    }

    // MARK: - Retrieval

    func getRecentLogs(lines: Int = 50) -> String {
        guard let data = FileManager.default.contents(atPath: logFile.path),
              let content = String(data: data, encoding: .utf8) else {
            return ""
        }
        let allLines = content.components(separatedBy: "\n")
        return allLines.suffix(lines).joined(separator: "\n")
    }

    func clearLogs() {
        try? "".write(to: logFile, atomically: true, encoding: .utf8)
    }

    // MARK: - File Management

    private func appendToFile(_ text: String) {
        if FileManager.default.fileExists(atPath: logFile.path) {
            // Rotate if too large
            if let attrs = try? FileManager.default.attributesOfItem(atPath: logFile.path),
               let size = attrs[.size] as? Int, size > maxLogSize {
                clearLogs()
            }
            if let handle = FileHandle(forWritingAtPath: logFile.path) {
                handle.seekToEndOfFile()
                handle.write(text.data(using: .utf8)!)
                handle.closeFile()
            }
        } else {
            try? text.write(to: logFile, atomically: true, encoding: .utf8)
        }
    }
}

File: shifai-android/app/src/main/java/com/shifai/data/monitoring/CrashReporter.kt
package com.shifai.data.monitoring

import android.content.Context
import android.util.Log
import com.shifai.data.analytics.AnalyticsTracker
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

/**
 * Crash Reporter — lightweight error tracking for production.
 * No third-party dependency: logs to local file + Plausible events.
 * Follows zero-PII principle.
 * Mirrors iOS CrashReporter.swift.
 */
object CrashReporter {

    private const val LOG_FILE = "crash_log.txt"
    private const val MAX_LOG_SIZE = 500_000L  // 500 KB
    private const val TAG = "CrashReporter"

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.FRANCE)
    private lateinit var logFile: File
    private var analyticsTracker: AnalyticsTracker? = null

    fun init(context: Context, tracker: AnalyticsTracker? = null) {
        logFile = File(context.filesDir, LOG_FILE)
        analyticsTracker = tracker

        // Set global uncaught exception handler
        val default = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            logCrash(throwable, "UncaughtException on ${thread.name}")
            default?.uncaughtException(thread, throwable)
        }
    }

    // ─── Logging ───

    fun log(errorCode: String, message: String, context: String = "") {
        val timestamp = dateFormat.format(Date())
        val line = "[$timestamp] [$errorCode] $message | $context\n"
        appendToFile(line)
        Log.w(TAG, line.trim())

        analyticsTracker?.let { tracker ->
            CoroutineScope(Dispatchers.IO).launch {
                tracker.trackError(errorCode)
            }
        }
    }

    fun logCrash(error: Throwable, context: String = "") {
        val timestamp = dateFormat.format(Date())
        val entry = "[CRASH] $timestamp | ${error::class.simpleName}: ${error.message} | $context\n"
        appendToFile(entry)
        Log.e(TAG, entry.trim())
    }

    // ─── Retrieval ───

    fun getRecentLogs(lines: Int = 50): String {
        if (!::logFile.isInitialized || !logFile.exists()) return ""
        val allLines = logFile.readLines()
        return allLines.takeLast(lines).joinToString("\n")
    }

    fun clearLogs() {
        if (::logFile.isInitialized) {
            logFile.writeText("")
        }
    }

    // ─── File Management ───

    private fun appendToFile(text: String) {
        try {
            if (!::logFile.isInitialized) return

            // Rotate if too large
            if (logFile.exists() && logFile.length() > MAX_LOG_SIZE) {
                clearLogs()
            }
            logFile.appendText(text)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to write log: ${e.message}")
        }
    }
}
